metronome/networking/src/io/iohk/metronome/networking/RemoteConnectionManager.scala

```diff
@@ -0,0 +1,339 @@
+        maxJitter =>
+          val randomJitterInMillis =
+            ThreadLocalRandom.current().nextLong(maxJitter.toMillis)
+          FiniteDuration(randomJitterInMillis, TimeUnit.MILLISECONDS)
+      }
+
+    val newDelay =
+      ((config.initialDelay * exponentialBackoff)
+        .min(config.maxDelay)) + randomJitter
+
+    Timer[F]
+      .sleep(newDelay)
+      .map(_ =>
+        failedConnectionRequest.copy(numberOfFailures = updatedFailureCount)
+      )
+  }
+
+  /** Connections are acquired in linear fashion i.e there can be at most one concurrent call to remote peer.
+    * In case of failure each connection will be retried infinite number of times with exponential backoff between
+    * each call.
+    */
+  private def acquireConnections[
+      F[_]: Concurrent: TaskLift: TaskLike: Timer,
+      K: Codec,
+      M: Codec
+  ](
+      encryptedConnectionProvider: EncryptedConnectionProvider[F, K, M],
+      connectionsToAcquire: ConcurrentQueue[F, OutGoingConnectionRequest[K]],
+      connectionsHandler: ConnectionHandler[F, K, M],
+      retryConfig: RetryConfig
+  ): F[Unit] = {
+
+    /** Observable is used here as streaming primitive as it has richer api than Iterant and have mapParallelUnorderedF
+      * combinator, which makes it possible to have multiple concurrent retry timers, which are cancelled when whole
+      * outer stream is cancelled
+      */
+    Observable
+      .repeatEvalF(connectionsToAcquire.poll)
+      .filterEvalF(request => connectionsHandler.isNewConnection(request.key))
+      .mapEvalF { connectionToAcquire =>
+        connectTo(encryptedConnectionProvider, connectionToAcquire)
+      }
+      .mapParallelUnorderedF(Integer.MAX_VALUE) {
+        case Left(failure) =>
+          //TODO add logging of failure
+          val failureToLog = failure.err
+          retryConnection(retryConfig, failure.connectionRequest).flatMap(
+            updatedRequest => connectionsToAcquire.offer(updatedRequest)
+          )
+        case Right(connection) =>
+          val newOutgoingConnections =
+            HandledConnection.outgoing(connection.encryptedConnection)
+          connectionsHandler.registerIfAbsent(newOutgoingConnections)
+
+      }
+      .completedF
+  }
+
+  /** Reads incoming connections in linear fashion and check if they are on cluster allowed list.
+    */
+  private def handleServerConnections[F[_]: Concurrent: TaskLift, K, M: Codec](
+      pg: EncryptedConnectionProvider[F, K, M],
+      connectionsHandler: ConnectionHandler[F, K, M],
+      clusterConfig: ClusterConfig[K]
+  ): F[Unit] = {
+    Iterant
+      .repeatEvalF(pg.incomingConnection)
+      .takeWhile(_.isDefined)
+      .map(_.get)
+      .collect { case Right(value) =>
+        value
+      }
+      .mapEval { encryptedConnection =>
+        clusterConfig.getIncomingConnectionServerInfo(
+          encryptedConnection.remotePeerInfo._1
+        ) match {
+          case Some(incomingConnectionServerAddress) =>
```

Comment:
Could this address be different from the one in the `encryptedConnection`? Maybe we should use that, since that is a fact, while the one in the `clusterConfig` could be misconfigured?

Reply:
I understand the reason now from your other comment :+1: